src/main/java/com/example/pokebattlez/battle/controller/Battle.java

```diff
@@ -22,45 +26,98 @@
-                .priority(activePokemon.get(trainerId).getStats().get(Stat.SPEED).getValue())
-                .build()
-        );
+    public void registerUseMove(Long trainerId, int movePosition, Long moveTarget) {
+        if (waitingForTrainers.contains(trainerId)) {
+            playerActions.add(PlayerAction.builder()
+                    .trainerId(trainerId)
+                    .moveInPosition(movePosition)
+                    .moveTarget(moveTarget)
+                    .action(Battle::useMove)
+                    .priority(activePokemon.get(trainerId).getStats().get(Stat.SPEED).getValue())
+                    .build()
+            );
+            waitingForTrainers.remove(trainerId);
+        }
     }
 
     public void cancelAction(Long trainerId) {
-        playerActions.stream()
-                .filter(playerAction -> playerAction.trainerId.equals(trainerId)).findFirst()
-                .ifPresent(playerActions::remove);
+        if (trainers.contains(trainerId) && !waitingForTrainers.contains(trainerId)) {
+            playerActions.stream()
+                    .filter(playerAction -> playerAction.trainerId.equals(trainerId)).findFirst()
+                    .ifPresent(playerActions::remove);
+            waitingForTrainers.add(trainerId);
+        }
     }
 
     private static void switchPokemon(Battle battle, PlayerAction playerAction) {
+        battle.turnLog.add(String.format("Called back %s!",
+                battle.activePokemon.get(playerAction.trainerId).getName()
+        ));
+
         battle.pokemon.get(playerAction.trainerId).stream()
                 .filter(pokemon1 -> pokemon1.getPosition() == playerAction.switchToPosition).findFirst()
                 .ifPresent(pokemon1 -> battle.activePokemon.replace(playerAction.trainerId, pokemon1));
+
+        battle.turnLog.add(String.format("Sent out %s!",
+                battle.activePokemon.get(playerAction.trainerId).getName()
+        ));
+
         battle.playerActions.remove(playerAction);
     }
 
     private static void useMove(Battle battle, PlayerAction playerAction) {
-        // TODO
+        Pokemon attacker = battle.activePokemon.get(playerAction.trainerId);
+        Pokemon defender = battle.activePokemon.values().stream().filter(pokemon1 -> pokemon1.getId().equals(playerAction.moveTarget)).findFirst().orElse(null);
+
+        Pokemon.Move move = attacker.getMoves().stream().filter(move1 -> move1.getPosition() == playerAction.moveInPosition).findFirst().orElse(null);
+
+        if (defender != null && move != null) {
+            int damage = battle.calculateDamage(attacker, defender, move);
+
+            defender.setCurrentHp(defender.getCurrentHp() - damage);
+
+            battle.turnLog.add(String.format(MOVE_USED_LOG,
+                    attacker.getName(),
+                    move.getName()
+            ));
+
+            int typeModifierIndex = battleUtils.getTypeModifierIndex(move.getType(), defender.getTypes());
```

Comment:
Is this only for informational reasons? Or maybe should be part of the calculateDamage mechanic?

Reply:
It's used both in calculateDamage and for informing the user about efficiency. I'm not worried about performance with this one since all it does is retrieve a value from a HashMap.